### PR TITLE
Adds a `normalizeSQLConstraint` method

### PR DESCRIPTION
**Pull request by mcdappdev**
Adds a `normalizeSQLConstraint` method to `SQLDialect` so that the drivers can truncate constraint identifiers that are too long.

### CODE REVIEW

File: Sources/SQLKit/Query/SQLConstraint.swift [hunk 1]
```diff
@@ -17,9 +17,23 @@ public struct SQLConstraint: SQLExpression {
...
             serializer.write(" ")
         }
         self.algorithm.serialize(to: &serializer)
     }
 }
+
+extension SQLDialect {
+    public func normalizeSQLConstraintIdentifier(_ identifier: String) -> String {
+        guard identifier.utf8.count > self.maximumConstraintIdentifierLength else { return identifier }
+        let midPoint = (identifier.count >> 1) - ((identifier.utf8.count - self.maximumConstraintIdentifierLength) >> 1)
```
[the remark this answers is not in the document]
Comment (gwynne) [reply Sources/SQLKit/Query/SQLConstraint.swift:35]: Isn't working with strings "correctly" just a ton of fun?

[the remark this answers is not in the document]
Comment (gwynne) [reply Sources/SQLKit/Query/SQLConstraint.swift:35]: Oh, and add tests which include complex Unicode characters in the test strings, such as `"smileyIdentifier😀IsASmiley"` (simple non-BMP emoji), `"ohWellWhatA🤷🏻‍♀️KindOfDay"` (complex variation selector emoji), and `"checkOutHowĈôm̂b̂în̂în̂ĝT⃠h⃠i⃠n⃠g⃠s⃠Works"` (combining characters on code points, which is multiple codepoints but only one grapheme cluster).

**Comment by gwynne** [comment]
Note: None of the drivers were updated to provide their own values for the `maximumConstraintIdentifierLength` property on `SQLDialect` because:

- **Postgres**: Limit is `63`, same as the default (since the algorithm conservatively assumes length - 1)
- **MySQL**: Limit is `64`, same as the default
- **SQLite**: Has no limit, so have it use the default to prevent surprising behavior for those switching drivers after starting with SQLite.
- **Mongo**: Not an SQL database anyway.

File: Sources/SQLKit/Query/SQLConstraint.swift [hunk 2]
```diff
@@ -17,9 +17,27 @@ public struct SQLConstraint: SQLExpression {
...
+                name.serialize(to: &serializer)
+            }
             serializer.write(" ")
         }
         self.algorithm.serialize(to: &serializer)
     }
 }
+
+extension SQLDialect {
+    public func normalizeSQLConstraintIdentifier(_ identifier: String) -> String {
```
Comment (tanner0101) [inline Sources/SQLKit/Query/SQLConstraint.swift:33]: Should this be a protocol requirement with a default implementation so that the dialect can choose what it wants to do?

File: Tests/SQLKitTests/SQLKitTests.swift [hunk 3]
```diff
@@ -259,6 +259,17 @@ CREATE TABLE `planets`(`id` BIGINT PRIMARY KEY AUTOINCREMENT, `name` TEXT DEFAUL
 """
                        )
     }
+    
+    func testConstraintLengthNormalization() {
+        XCTAssertEqual(
+            db.dialect.normalizeSQLConstraintIdentifier("fk:obnoxiously_long_table_name.other_table_name_id+other_table_name.id"),
```
Comment (tanner0101) [inline Tests/SQLKitTests/SQLKitTests.swift:265]: Maybe we should put a `...` in the middle so the truncation is clearer? 

```
// too long
fk:foo.bar+baz.quux
// truncates to
fk:fo...ux
```

File: Sources/SQLKit/Query/SQLConstraint.swift [hunk 4]
```diff
@@ -17,7 +17,12 @@ public struct SQLConstraint: SQLExpression {
     public func serialize(to serializer: inout SQLSerializer) {
         if let name = self.name {
             serializer.write("CONSTRAINT ")
-            name.serialize(to: &serializer)
+            if let identifier = (name as? SQLIdentifier)?.string {
```
Comment (tanner0101) [inline Sources/SQLKit/Query/SQLConstraint.swift:20]: Casting as `SQLIdentifier` here could cause problems.

File: Sources/SQLKit/SQLDialect.swift [hunk 5]
```diff
@@ -14,6 +14,7 @@ public protocol SQLDialect {
     var triggerSyntax: SQLTriggerSyntax { get }
     var alterTableSyntax: SQLAlterTableSyntax { get }
     func customDataType(for dataType: SQLDataType) -> SQLExpression?
+    func normalizeSQLConstraintIdentifier(identifier: String) -> String
```
Comment (tanner0101) [inline Sources/SQLKit/SQLDialect.swift:17]: This should accept and return `SQLExpression`.

Comment (tanner0101) [reply Sources/SQLKit/SQLDialect.swift:17]: `identifier` is mentioned twice here. I'd use either:
```swift
normalizeSQLConstraintIdentifier(_:)
// or
normalizeSQLConstraint(identifier:)
```

**Comment by tanner0101** [comment]
These changes are now available in [3.0.0-rc.2](https://github.com/vapor/sql-kit/releases/tag/3.0.0-rc.2)
